jenkins/README.md

```diff
@@ -0,0 +1,157 @@
+# CI instrumentation scripts
+Scripts under this directory are used for deploying tests in the CI. These are kept with source for following reasons:
+
+- Ease of defining and maintaining different CI jobs.
+- Reduce CI script complexity.
+- Revision specific testing.
+- Code reusability.
+- Quality control via reviews and possibly testing in future.
+- Debug CI scripts on host.
+
+## CI Infrastructure and test entities relation
+Following entity relation diagram shows the relation between different components involved in Mbed TLS testing:
+```
+              --------------- m        1 ----------
+              | Slave label |------------|   CI   |
+              ---------------            ----------
+                      |1                      | 1
+                      |                       |
+                      |                       | m
+                      |                  ----------
+                      |                  |  Job   |
+                      |                  ----------
+                      |                       | m
+                      |                       |
+                      |1                      | m
+                 ------------ m      m --------------- 1      m --------------------
+                 | Platform |----------| Test script |----------| Test environment |
+                 ------------          ---------------          --------------------
+
+```
+Terms used in the above diagram
+- **CI** Continuous Integration infrastructure.
+- **Slave label** Label identifying a type of slave machines commissioned for running specific tests.
+- **Job** A test campaign run on a particular revision of source. Like PR, nightly, release testing.
+- **Test script** A test script or bunch of commands. It is executed on each required platform and environment.
+- **Platform** Target platform. Like Ubuntu, Debian, FreeBSD etc.
+- **Test environment** Environment to chang a test script behaviour. Like environment ```CC=gcc``` selects compiler GCC.
+
+## CI meta data
+Based on above entity relation CI jobs can be defined with the meta data in following json format:
+
+```py
+ci_jobs = {
+   "commit_tests": {
+       'make-gcc': {
```

Comment:
> 1. Configure: would typically either copy a file from configs or would run config.pl zero or more times to modify the default config.h.
> 2. Build: select between make/cmake/VS/… and also select CC, CFLAGS, etc.
> 3. Test: whether to run make check, whether to run ssl-opt.sh and compat.sh and with what options, etc.

Would following data format look good to you:
```
{'config': 'full', 'tests': ['full'], 'build': {'name': 'cmake-asan', 'environment': {'CC': 'clang'}}}
```
On above I would like to suggest to move ```environment``` one level up at same level as other attributes. It will give us a flexible interface to pass additional variables in future. Hence it would look like:

```
{'config': 'full', 'tests': 'full', 'build': 'cmake-asan', 'environment': {'CC': 'clang'}}
```

Sorry the README is not up to date. The implementation is nearly same. Only difference is that config is passed as an environment variable. I will update code after your approval.

> Note that in addition to configure-build-test jobs, there are other jobs, e.g. to run check-*.sh or yotta.

We can make ```config``` and ```tests``` optional and ```build``` can specify a script with path. ```ciscript.sh/bat``` will understand this and run the script. Again any environment requirement for the script will be passed via ```environment``` attribute.

Coming back to this point:
> Rather than specifying a job globally, I would like a compositional way to describe a job as a configure-build-test cycle.

The name specified by the user simplifies the scripts. We need name for a composition to name the CI parallel branches, name logs, putting in the report. The scripts will try to form the name based on the selected config, test, build, environment and it could get big and less user identifiable. But if a user specifies the name it can be based on the intent. Like: cmake-asan-full where it is assumed the that compiler is ```clang``` and all the tests should be run. Let me know what you think.

Note: I will update the readme once we have closure on the design. So that this discussion is not lost.

Reply:
The data format you're proposing looks fine. I'm not too concerned about the detail of the structure, it's mostly about what information you can specify.

> We can make config and tests optional and build can specify a script with path. 

I think it would make more sense for `tests` to be mandatory and `build` optional, but that's a detail.

> We need name for a composition to name the CI parallel branches, name logs, putting in the report.

You can build a compound name, e.g. `full+make-gcc+check` which `ciscript` parses into “`full` config, `make-gcc` build, `check` test”.

Reply:
> I think it would make more sense for tests to be mandatory and build optional, but that's a detail.

Ok so when we have scripts we put them as a test rather than a build.

If you have observed I have made ```tests``` as a list. So that we can specify a list of tests. Currently it is allowed to have 0 tests. But I will make it have at list one. The test would either be script to run or a name that ``ciscript.sh/bat``` translates into bunch of commands.

Reply:
As discussed offline we can use names for defined subjobs in ```cijobs.json```. Any other script that consumes ```cijobs.json``` will be generic and depend on data provided by ```cijobs.json```. This will help simplify CI that needs names for parallel execution branches, logs and reports.